### PR TITLE
[Bugfix][HunyuanImage3] Pin AR ratio token to requested size

### PR DESCRIPTION
**Pull request by TaffyOfficial**
## Purpose

HunyuanImage3 uses an AR stage to emit image control tokens before DiT generation. For explicit-size requests, the serving path already carries the requested `height` / `width` as `target_h` / `target_w`, and downstream DiT sizing expects the AR tokens to agree with that requested bucket.

The missing piece is the AR ratio-token step. After `<img_size_*>`, current `main` still lets `_apply_ratio_restriction()` greedily choose the highest-scoring `<img_ratio_*>` token. That means a request with an explicit size such as `1280x768` can still have AR select a different ratio bucket, so the user-requested size is no longer authoritative.

This PR makes explicit size win at the AR sampler boundary:

- store per-request `sampling_extra_args` during AR forward
- resolve `target_h` / `target_w` through the HunyuanImage3 resolution table
- force the matching `<img_ratio_*>` token by masking all other ratio logits
- keep the existing greedy ratio selection when no explicit target size is present

This does not change condition-image forwarding between stages. It only uses the explicit target-size metadata that is already attached to the AR request, so AR ratio selection, AR stop-token behavior, and DiT sizing stay on the same requested bucket.

## Test Plan

Coverage in this PR:

- forced ratio-token selection when `target_h` / `target_w` is present, including the case where the matching ratio token has a lower logit than the greedy token
- existing greedy ratio fallback when no explicit target size is present
- existing extra ratio-token slice behavior
- serving path sets AR stop-token constraints for complete explicit size requests
- partial explicit size requests are rejected instead of creating inconsistent AR/DiT state
- no-size requests do not get explicit-size stop-token constraints
- multi-stage online image-editing path still forwards the condition images into the downstream DiT stage

## Test Result

Current PR head: `d8530483a05d267334446bfaccb61b622c9a84a5`.

Focused sampler regression:

- `tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py::TestRatioRestriction::test_target_size_forces_matching_ratio_token`: `1 passed`

Multi-stage online accuracy check:

- Base: `d897852258d9bcc05101c54dcd0973d111de9723`
- PR head: `d8530483a05d267334446bfaccb61b622c9a84a5`
- Workload: `tests.e2e.accuracy.test_hunyuan_image3._run_online`
- API path: `/v1/images/edits`
- Request: `size=1280x720`, seed `42`, `50` inference steps, guidance scale `2.5`
- Before/after output: https://github.com/vllm-project/vllm-omni/pull/4501#issuecomment-4726580717

| Metric | Base | PR head |
| --- | ---: | ---: |
| Output size | `1280x720` | `1280x720` |
| COT prefix match | `127` | `127` |
| Required COT markers present | yes | yes |
| Image-Image similarity | `97.2845` | `97.2845` |
| SSIM | `0.2722` | `0.2722` |
| PSNR | `14.19 dB` | `14.19 dB` |

The base and PR output PNGs are byte-identical for this multi-stage case (`sha256=29230ca372fdc612bc226defa20c7fd59d05200daa64b6f739b8a4a49f4b700e`; pixel diff is empty), so this PR does not regress condition-image handoff into the downstream DiT stage.

### CODE REVIEW

**Comment by TaffyOfficial** [comment]
@Bounty-hunter

**Comment by Bounty-hunter** [comment]
It seem that in muti-stage case, the following stage can't get the input image, and will result accuracy problem. can you show the case and output before and after this pr?

**Comment by TaffyOfficial** [comment]
> It seem that in muti-stage case, the following stage can't get the input image, and will result accuracy problem. can you show the case and output before and after this pr?

I reran the multi-stage online image-editing case on the PR base and on the current PR head.

Case:
- Base: `d897852258d9bcc05101c54dcd0973d111de9723`
- PR head: `d8530483a05d267334446bfaccb61b622c9a84a5`
- Workload: `tests.e2e.accuracy.test_hunyuan_image3._run_online`
- API path: `/v1/images/edits`
- Request: `size=1280x720`, seed `42`, `50` inference steps, guidance scale `2.5`

Before / after output:

PR4501 before after online accuracy

Accuracy / consistency result:

| Metric | Base | PR head |
| --- | ---: | ---: |
| Output size | `1280x720` | `1280x720` |
| COT prefix match | `127` | `127` |
| Required COT markers present | yes | yes |
| Image-Image similarity | `97.2845` | `97.2845` |
| SSIM | `0.2722` | `0.2722` |
| PSNR | `14.19 dB` | `14.19 dB` |

The base and PR output PNGs are byte-identical in this case (`sha256=29230ca372fdc612bc226defa20c7fd59d05200daa64b6f739b8a4a49f4b700e`; pixel diff is empty). So this PR does not change or break the condition-image handoff to the following stage in this multi-stage path. The code change is limited to forcing the AR ratio token to match explicit `target_h` / `target_w`; condition-image forwarding remains unchanged.

I also reran the focused sampler regression on the current PR head: `tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py::TestRatioRestriction::test_target_size_forces_matching_ratio_token` passed (`1 passed`).

**Review by hsliuustc0106 — COMMENTED** [review]
Bugfix looks correct.

**Comment by zengchuang-hw** [comment]
LGTM

**Comment by zengchuang-hw** [comment]
LGTM

**Comment by Bounty-hunter** [comment]
It looks like the current main branch can produce an incorrect output size when an explicit width/height is specified.

Could you share a repro case? I couldn't find where `sampling_extra_args.target_w/target_h` are set in `https://github.com/vllm-project/vllm-omni/blob/main/examples/offline_inference/hunyuan_image3/end2end.py`, so I'm not sure how to trigger this path.

**Comment by TaffyOfficial** [comment]
> It looks like the current main branch can produce an incorrect output size when an explicit width/height is specified.
> 
> Could you share a repro case? I couldn't find where `sampling_extra_args.target_w/target_h` are set in `https://github.com/vllm-project/vllm-omni/blob/main/examples/offline_inference/hunyuan_image3/end2end.py`, so I'm not sure how to trigger this path.

Good catch. The first commit fixed the consumer side: once AR receives `sampling_extra_args.target_h/target_w`, it forces the matching ratio token.

There are now both online and offline repro paths.

Online `/v1/images/generations`:

```bash
curl http://localhost:8000/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "prompt": "A cute cat",
    "size": "1280x768",
    "bot_task": "think"
  }'
Online /v1/images/edits with explicit size, not size=auto:
curl http://localhost:8000/v1/images/edits \
  -F prompt="Make this image into a fridge magnet" \
  -F image="@input.png" \
  -F size="1280x768" \
  -F bot_task="think"
Offline:
python examples/offline_inference/hunyuan_image3/end2end.py \
  --modality text2img \
  --prompts "A cute cat" \
  --width 1280 \
  --height 768 \
  --bot-task think
For online generation, this is already covered by tests/entrypoints/openai_api/test_image_server.py::test_generate_images_async_omni_glm_image_sets_stage0_max_tokens, which checks that stage-0 AR sampling params receive:
extra_args["target_h"] = 1024
extra_args["target_w"] = 1024
